system/lib/wasmfs/backends/ignore_case_backend.cpp

```diff
@@ -20,105 +37,171 @@ std::string normalize(const std::string& name) {
   }
   return result;
 }
-} // namespace
+
+} // anonymous namespace
 
 namespace wasmfs {
 
-// Problem: Child entries are stored both in IgnoreCaseDirectory and in baseDirectory.
-//
-// Original name case preservation could be possible if MemoryDirectory::entries
-// were accesible. Then, we would store the original case there but search
-// ignoring case.
-//
-class IgnoreCaseDirectory : public MemoryDirectory {
-  using BaseClass = MemoryDirectory;
-  std::shared_ptr<Directory> baseDirectory;
+class IgnoreCaseDirectory : public VirtualDirectory {
 
-public:
-  IgnoreCaseDirectory(std::shared_ptr<Directory> base, backend_t proxyBackend)
-    : BaseClass(base->locked().getMode(), proxyBackend), baseDirectory(base) {}
+  struct ChildInfo {
+    std::string originalName;
+    std::shared_ptr<File> child;
+  };
 
-  std::shared_ptr<File> getChild(const std::string& name) override {
-    return BaseClass::getChild(normalize(name));
-  }
+  // Map normalized names to virtual files and their non-normalized names.
+  std::map<std::string, ChildInfo> children;
 
+public:
+  IgnoreCaseDirectory(std::shared_ptr<Directory> real, backend_t backend)
+    : VirtualDirectory(real, backend) {}
+
+  std::shared_ptr<File> getChild(const std::string& name) override;
```

Comment:
What is the logic to define something in class vs. separate declaration/definiton? Anything except 1-liners should be split?

Reply:
Yeah, splitting out nontrivial methods is usually a good idea, but less important in .cpp files, I think. In this case, the `virtualize` implementations depended on the full class definitions and the method definitions depended on `virtualize`, so I had to either forward declare the `virtualize` functions or split out the method definitions. I chose the latter.